docs/design-proposals/high-availability.md

```diff
@@ -0,0 +1,209 @@
+
+#### Runtime
+
+Make runtime attempt to re-connect to the proxy if it is unable to
+connect initially. This requires changes to the `virtcontainers`
+`ccProxy.connectProxy()` function to make it:
+
+- Reconnect and timeout after a period (like the [shim reconnect PR]).
+
+- Record details of the last command `virtcontainers` attempted to send to the
+  proxy (`hyperstartProxyCmd`) in memory.
+
+- If a proxy reconnect is required, re-send the command.
+
+#### Proxy
+
+No changes.
+
+#### Shim
+
+Make the shim cache all signals, input data and the last command run and
+re-send to the proxy when it reconnects after being disconnected from
+the proxy.
+
+#### Agent
+
+- Handle gracefully the re-connection of the proxy to continue properly
+where we left.
+
+- Buffer all outputs supposed to go through STDOUT/STDERR, so that the
+agent can send them after the proxy re-connection. Specific to IO
+channel.
+
+- Save the last command that got executed while the proxy was crashing,
+and save the result. The idea is that when the proxy is gonna reconnect,
+it's gonna send again the last command because it didn't get the result
+(this command is really gonna be triggered by the shim or the runtime
+when they reconnect). For that reason, the agent should analyze the
+command sent by the proxy after it reconnects, and not execute it (in
+case that matches the last command), but send the saved result, to avoid
+running the same command a second time. That way, the proxy will receive
+the result of this command.
+
+- Always save the last outputs that we are sending. This would allow the
+agent to resend the result when the same command is submitted from the
+shim or runtime, but that we don't want to re-run the command for real,
+because it could have different results.
+
+- Modify the agent service to restart on failure once it is able to
+  reconnect to the workload and proxy.
+
+#### Workload
+
+#### Hypervisor
+
+Identify how the current behaviour can be improved; if the proxy
+currently stops, the hypervisor will be left running consuming a large
+amount of CPU due to the agent attempting to reconnect to the proxy.
+The reconnect behaviour is correct, but there is no timeout in the case
+where the proxy needs to be manually stopped by an administrator for example.
+
+### Required test changes
+
+#### Mock components
+
+Mock components and frameworks will need to be updated to match the new behaviour above.
+
+#### Scenario testing
+
+To ensure every failure scenario is covered, we need to provide a way to
+stimulate all types of major failure. This needs further exploration but
+we need to consider whether:
+
+- it may be possible provoke some/most scenarios using our mock systems.
+
+- we can use other tooling to for example guarantee no available disk
+  space / memory.
+
+- potentially we need to change the actual components to provide a way
+  to make them fail "on demand".
+
+
+### Scenarios that need testing
+
```

Comment:
Suggested introductory line: Scenarios that need testing consist of disconnects, `ENOSPC`, `ENOMEM`, limits, and logging.